flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/FlinkKafkaInternalProducer.java

```diff
@@ -74,6 +76,10 @@ private static Properties withTransactionalId(
 
     @Override
     public Future<RecordMetadata> send(ProducerRecord<K, V> record, Callback callback) {
+        if (flushGated) {
```

Comment:
I don't think this helps because during a SinkWriter#flush(), a SinkWriter#write() cannot be invoked. So this boolean condition never occurs. For your issue, we are looking to replicate this behavior https://github.com/apache/flink-connector-kafka/blob/15f2662eccf461d9d539ed87a78c9851cd17fa43/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer.java#L1107 to ensure that all records are flushed before completing the checkpoint, even in the case of Kafka client timeouts/retries.

Reply:
reworked to follow [flink-connector-kafka/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer.java](https://github.com/apache/flink-connector-kafka/blob/15f2662eccf461d9d539ed87a78c9851cd17fa43/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducer.java#L1107)